synapse/lib/node.py

```diff
@@ -382,4 +356,64 @@ def delTag(self, tag, init=False):
         sops.append(('prop:del', (self.buid, self.form.name, '#' + name, info)))
 
         self.snap.stor(sops)
-        self.snap.splice('tag:add', ndef=self.ndef, tag=name, valu=name)
+        self.snap.splice('tag:del', ndef=self.ndef, tag=name, valu=curv)
+
+    def delete(self):
+        '''
+        Delete a node from the cortex.
+
+        The following tear-down operations occur in order:
+
+            * validate that you have permissions to delete the node
+            * validate that you have permissions to delete all tags
+            * validate that there are no remaining references to the node.
+
+            * delete all the tags (bottom up)
+                * fire onDelTag() handlers
+                * delete tag properties from storage
+                * log tag:del splices
+
+            * delete all secondary properties
+                * fire onDelProp handler
+                * delete secondary property from storage
+                * log prop:del splices
+
+            * delete the primary property
+                * fire onDel handlers for the node
+                * delete primary property from storage
+                * log node:del splices
+        '''
+
+        formname, formvalu = self.ndef
+
+        # check permissions
+        if not self.snap.allowed('node:del', formname):
+            return self.snap._onAuthDeny('Not allowed to delete the node.')
+
+        tags = [(len(t), t) for t in self.tags.keys()]
+
+        # check for tag permissions
+        for size, tag in tags:
+            tagpath = s_chop.tagpath(tag)
+            if not self.snap.allowed('tag:del', *tagpath):
+                return self.snap._onAuthDeny('Not allowed to delete node with tag {tag}.')
+
+        # check for any nodes which reference us...
+        if any(self.snap._getNodesByType(formname, formvalu, addform=False)):
```

Comment:
We should include unit test to ensure we can't end up in situation where two nodes that refer to each other can't be deleted